From fabbe2965c579883c21e52c76be2dc9dc2a3ce05 Mon Sep 17 00:00:00 2001
From: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
Date: Sat, 17 Sep 2016 01:35:29 -0700
Subject: [PATCH 4/5] os/connection: Call ClientReady() based on a level
 trigger rather than an edge trigger

On encountering an EAGAIN, FlushClient() re-adds the client to the pending list
and returns, but FlushClient() doesn't get called again.  We would expect it to
be called via FlushAllOutput() via WaitForSomething(), but that only happens
when NewOutputPending is set.  FlushAllOutput() also early-exits if
NewOutputPending is not set.

The only place that is setting NewOutputPending is ClientReady().  The problem
there is that ClientReady() is called based on an edge trigger and not a level
trigger.

Signed-off-by: Jeremy Huddleston Sequoia <jeremyhu@apple.com>
CC: Keith Packard <keithp@keithp.com>
---
 os/connection.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/os/connection.c b/os/connection.c
index 0d42184..4630438 100644
--- a/os/connection.c
+++ b/os/connection.c
@@ -749,7 +749,7 @@ AllocNewConnection(XtransConnInfo trans_conn, int fd, CARD32 conn_time)
     SetConnectionTranslation(fd, client->index);
 #endif
     ospoll_add(server_poll, fd,
-               ospoll_trigger_edge,
+               ospoll_trigger_level,
                ClientReady,
                client);
     set_poll_client(client);
-- 
2.10.0 (Apple Git-77)

